drivers/dai/nxp/sai.c

```diff
@@ -0,0 +1,758 @@
+
+	/* disable error interrupt */
+	SAI_TX_RX_ENABLE_DISABLE_IRQ(dir, data->regmap,
+				     kSAI_FIFOErrorInterruptEnable, false);
+
+	return 0;
+}
+
+static int sai_trigger_start(const struct device *dev,
+			     enum dai_dir dir)
+{
+	struct sai_data *data;
+	uint32_t old_state;
+	int ret;
+
+	data = dev->data;
+	old_state = sai_get_state(dir, data);
+
+	/* TX and RX should be triggered independently */
+	if (dir != DAI_DIR_RX && dir != DAI_DIR_TX) {
+		LOG_ERR("invalid direction: %d", dir);
+		return -EINVAL;
+	}
+
+	/* attempt to change state */
+	ret = sai_update_state(dir, data, DAI_STATE_RUNNING);
+	if (ret < 0) {
+		LOG_ERR("failed to transition to RUNNING from %d. Reason: %d",
+			sai_get_state(dir, data), ret);
+		return ret;
+	}
+
+	if (old_state == DAI_STATE_PAUSED) {
+		/* if the SAI has been paused then there's no
+		 * point in issuing a software reset. As such,
+		 * skip this part and go directly to the TX/RX
+		 * enablement.
+		 */
+		goto out_enable_tx_rx;
+	}
+
+	LOG_DBG("start on direction %d", dir);
+
+	if (dir == DAI_DIR_RX) {
+		/* this is fine because TX is async so it won't be
+		 * affected by an RX software reset.
+		 */
+		SAI_TX_RX_SW_RESET(dir, data->regmap);
+
+		/* do a TX software reset only if not already enabled */
+		if (!data->tx_enabled) {
+			SAI_TX_RX_SW_RESET(DAI_DIR_TX, data->regmap);
+		}
+	} else {
+		/* a software reset should be issued for TX
+		 * only if RX was not already enabled.
+		 */
+		if (!data->rx_enabled) {
+			SAI_TX_RX_SW_RESET(dir, data->regmap);
+		}
+	}
+
+	/* enable error interrupt */
+	SAI_TX_RX_ENABLE_DISABLE_IRQ(dir, data->regmap,
+				     kSAI_FIFOErrorInterruptEnable, true);
+
+	/* TODO: is there a need to write some words to the FIFO to avoid starvation? */
+
+	/* TODO: for now, only DMA mode is supported */
+	SAI_TX_RX_DMA_ENABLE_DISABLE(dir, data->regmap, true);
+
+	/* enable TX/RX data line. This translates to TX_DLINE0/RX_DLINE0
+	 * being enabled.
+	 *
+	 * TODO: for now we only support 1 data line per direction.
+	 */
+	sai_tx_rx_set_dline_mask(dir, data->regmap, 0x1);
+
+out_enable_tx_rx:
+	/* this will also enable the async side */
+	SAI_TX_RX_ENABLE_DISABLE(dir, data->regmap, true);
+
+	/* update the software state of TX/RX */
+	sai_tx_rx_sw_enable_disable(dir, data, true);
+
+	return 0;
+}
+
+static int sai_trigger(const struct device *dev,
+		       enum dai_dir dir,
+		       enum dai_trigger_cmd cmd)
+{
+	switch (cmd) {
+	case DAI_TRIGGER_START:
+		return sai_trigger_start(dev, dir);
+	case DAI_TRIGGER_PAUSE:
+		return sai_trigger_pause(dev, dir);
+	case DAI_TRIGGER_STOP:
+		return sai_trigger_stop(dev, dir);
+	case DAI_TRIGGER_PRE_START:
+	case DAI_TRIGGER_POST_STOP:
+	case DAI_TRIGGER_DRAIN:
+	case DAI_TRIGGER_DROP:
+	case DAI_TRIGGER_PREPARE:
+	case DAI_TRIGGER_RESET:
+	case DAI_TRIGGER_COPY:
+		/* TODO: maybe return an error code here
+		 * instead of making it seem like the
+		 * trigger operation has succeeded.
+		 */
```

Comment:
This may just end up polluting the logs. For example, `dai_common_copy()` does `DAI_TRIGGER_COPY` which is not needed for us. Will have to double-check and see if this is manageable. If it's not and you think this may be helpful we could go with a `LOG_DBG` instead?